compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1588,6 +1607,47 @@ impl Type {
         }
     }
 
+    // If `self` and `expected` are function types, tries to coerce `self` to `expected`.
+    // Returns None if this produced an error, otherwise returns Some(make_error) meaning
+    // that regular unification must still be done (and it could produce an error).
+    pub fn try_fn_to_unconstrained_fn_coercion(
+        &self,
+        expected: &Type,
+        expression: ExprId,
+        interner: &mut NodeInterner,
+        errors: &mut Vec<TypeCheckError>,
+        make_error: impl FnOnce() -> TypeCheckError,
+    ) -> Option<impl FnOnce() -> TypeCheckError> {
+        // If `self` and `expected` are function types, `self` can be coerced to `expected`
+        // if `self` is unconstrained and `expected` is not. The other way around is an error, though.
+        let (
+            Type::Function(params, ret, env, unconstrained_self),
+            Type::Function(_, _, _, unconstrained_expected),
+        ) = (self.follow_bindings(), expected.follow_bindings())
```

Comment:
We shouldn't match on a single variant of a type without calling `follow_bindings` first. Since otherwise the type may be a `Type::TypeVariable` bound to a `Type::Function` but would fail this check.

Reply:
Sorry if I'm misunderstanding, but there's `self.follow_bindings()` and `expected.follow_bindings()` here. Should they be called somewhere else too?

Reply:
Just the two of those should be sufficient. Those follow type variable links so any time we're matching on the variant of a type like the `let .. else` statement here, we need to call these methods first.

Reply:
Do you mean I should pattern match them but first one, then the other?

Reply:
I'm sorry, I'm not sure how I missed it but I somehow missed that we already call `follow_bindings()` before matching. Ignore me.

Reply:
All good! 😄